### PR TITLE
add organization update functionality for forms

### LINKED ISSUE #35: Handle form submissions that have null value protection

### PR DESCRIPTION
**Pull request by Sarah-Addo**

### CODE REVIEW

File: src/main/java/com/google/sps/data/Organization.java
Comment (andrewm898) [inline src/main/java/com/google/sps/data/Organization.java]: Where are these variables initialized / stored? it looks like multiple functions reference them but I don't see them included as class variables

edit: nevermind, github just collapsed that code. Should we be using the this. keyword on these though?

Comment (Sarah-Addo) [reply src/main/java/com/google/sps/data/Organization.java]: added this keyword

**Review by jnnybk — APPROVED** [review]

File: src/main/java/com/google/sps/data/Organization.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/Organization.java]: Why? Won't this clear out the organization description?

File: src/main/java/com/google/sps/data/Organization.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/Organization.java]: do you mean, this.name, this.email, etc?

File: src/main/java/com/google/sps/data/Organization.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/Organization.java]: or null?

File: src/main/java/com/google/sps/data/Organization.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/Organization.java]: instead of storing the data in separate variables as you do above (and the commit function where you set them directly), what about making these into functions?


example:
constructor() {
  this.entity = entity;
}

long id() {
   return (long) entity.getKey().getId();
}
String name() {
   return entity.getProperty("orgName");
}
etc

File: src/main/java/com/google/sps/data/Organization.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/Organization.java]: If you store the entity, this can then instead be:

this.entity_.setProperty(propertyKey, formValue);

for everything except moderatorList and isApproved

ie;

if (propertyKey is moderatorList) {
  // what you have now
} else if (propertyKey is apprived) {
  // what you have now
} else {
  // this handles everything else.
  this.entity_.setProperty(propertyKey, formValue);
}

File: src/main/java/com/google/sps/data/Organization.java
Comment (belraquib) [reply src/main/java/com/google/sps/data/Organization.java]: I think you don't need this function if you store just this.entity

File: src/main/java/com/google/sps/data/Organization.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/Organization.java]: instead of checking for isApproved, what if you store a Set?

Set<String> requiresMaintainer; requiresMaintainer.add("a");

same for requestModerator


Then this code checks if the propertyKey is in the maintainer/moderator set and it's easy to add new fields that require permissions.

**Review by belraquib — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/data/Organization.java
Comment (Sarah-Addo) [reply src/main/java/com/google/sps/data/Organization.java]: added why we are doing this in the design doc

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (mickrobk) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java]: What do you think of accepting a GiverUser instead of these booleans?

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (mickrobk) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java:75]: "or" here doesn't line up with this if statement

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java]: nit:  move to line 71

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (mickrobk) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java:110]: Is this right? Aren't you storing userids in the db, and accepting emails in the form?

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java]: this can be null. what should your code do if it is null? set it to an empty string?

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java]: check lint. spacing looks off

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (mickrobk) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java]: do you need the if on the else here? Or can you say it is not approved if "approved" doesn't match?

**Review by mickrobk — CHANGES_REQUESTED** [review]

**Review by belraquib — CHANGES_REQUESTED** [review]
This looks really good! My only concern is how you'd handle null values for the org description

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (belraquib) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java:110]: Good catch here.

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (belraquib) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java]: an alternative is to assume the JS side sends true/false  -  I wonder if you can extend Andrew's  CoerceToBool change (moving into a more generic RequestHelper class) to accept what the expected "true" param would be.

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (Sarah-Addo) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java]: So right now GiverUser does not include an IsModerator check or field so I would have to add stuff to GiverUser which is still a work in progress in order to test. Should I go ahead and do it anyways or add a TODO?

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (Sarah-Addo) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java:75]: I think it does because it would do the continue if a request was not sent by a moderator or a maintainer. The check would fail if it was a moderator or then fail if it was a maintainer. I think the key here is that the isModerator check is before the isMaintainer check.

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (Sarah-Addo) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java:110]: Now will be using givr user translate email to ids function

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (mickrobk) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java:75]: OK, I think you're right. Is there a way to rewrite this so there are less "nots"?
Eg currently this is
allow access = not(a && not(b) && not(c))

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (mickrobk) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java]: There is currently a GiverUser checked in right; what are your concerns here? Is there a middle ground option?

Comment (mickrobk) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java]: To be clear: the heart of my question is basically: if we have this abstraction for "User," why aren't we using it and extending it? I want to make sure the work we do complements each other; that we're not just pulling in different directions.

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (Sarah-Addo) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java:75]: yes i think its currently a && !b && !c so it would be  a && not( b || c) by demorgans law

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (Sarah-Addo) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java]: My concern is that at the moment there GivrUser does not have a way of fulfilling the isModerator boolean check. On the edit org servlet side it would pass in request, GivrUser.isMaintainer, true since the edit functionality would only be displayed to users who are at least moderators in the first place. But when it comes from the Add Organization side it does not care who the user is since the registration form only takes care of certain params which is why it calls it with request, false, false. If I replaced this with a GiverUser then isMaintainer and isModerator would both have to result in false in order for the isApproved and ModeratorList params to not be checked which could be false since they could be a maintainer and they would technically be the first moderator for the org that they are submitting the request for. The way im looking at it now these checks are more for specifying which servlet is calling this function. I am thinking that if we did pass in a GiverUser param then we would also have to pass in if its calling update for registering or not so it ignores the user which seems like what makes the most sense

Comment (mickrobk) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java]: It seems like an error prone pattern to ask the caller of this function to sometimes lie about the authorization level of the user! Is there another way we could solve this?

Comment (belraquib) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java]: Consider: right now you are planning on sending down just "true" for isModerator right?

If we assume GIvrUser will do this can you can store this as local variables. For now you can always set isModerator to false.

On the consideration of different behavior for the caller: you should be careful about using  isModerator/isMaintainer to signify who the caller is. If you really have to change behavior based on who is calling the function, then consider a different parameter. If you did want to continue to use the call as if the caller is always unauthenticated, then you can create a subclass of the GivrUser or somehow consruct a GivrUser that guarantees it has no permissions.

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java]: this will be a bit simpler to read if you used some helper vars;


bool propertyRequiresMaintainer = requiresMaintainer.contains(propertyKey);
bool propertyRequiresModerator = requiresModerator.contains(propertyKey);

Then the rest of the code will read a bit easier (and you won't recompute the set membership)

Comment (belraquib) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java]: I would even add an additional:

bool propertyRequiresAuth = propertyRequiresMaintainer || propertyRequiresModerator;

now this condition becomes

if (forRegistration && propertyRequiresAuth) {
  continue;
}

which really makes it easy to read

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (belraquib) [inline src/main/java/com/google/sps/data/OrganizationUpdater.java:127]: can the user returned ever be null? or can the email address fail to generate a user id (from a typo, for instance!).

**Review by belraquib — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/data/OrganizationUpdater.java
Comment (Sarah-Addo) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java:127]: reading the documentation it seems that it will always make a new user even if a typo exists so that unique id will be added to the list

Comment (belraquib) [reply src/main/java/com/google/sps/data/OrganizationUpdater.java:127]: Looks like it can return null

https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/users/User.html#getuserid

Reading that, I'm not sure if we can get the user from an email address. Can you file an issue to investigate?

**Review by belraquib — APPROVED** [review]

**Pull request merged**
